File: backend/app/schemas/foreshadow.py
"""Strict API contracts for durable foreshadow planning and facts."""

from __future__ import annotations

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator


ForeshadowState = Literal["unplanted", "planted", "pending_resolution", "resolved"]


class ForeshadowOperationCommand(BaseModel):
    operation_key: str = Field(
        min_length=8,
        max_length=128,
        pattern=r"^[A-Za-z0-9._:-]+$",
    )

    model_config = ConfigDict(extra="forbid")


class ForeshadowBindCommand(ForeshadowOperationCommand):
    element_id: str = Field(min_length=32, max_length=32)
    expected_structure_version: int = Field(ge=1)
    expected_element_lock_version: int = Field(ge=1)


class ForeshadowLifecycleCommand(ForeshadowOperationCommand):
    expected_lifecycle_version: int = Field(ge=1)


class ForeshadowRestoreCommand(ForeshadowLifecycleCommand):
    expected_structure_version: int = Field(ge=1)
    expected_element_lock_version: int = Field(ge=1)


class ForeshadowPlanCreate(ForeshadowLifecycleCommand):
    expected_structure_version: int = Field(ge=1)
    action_kind: Literal["plant", "resolve"]
    target_type: Literal["part", "chapter"]
    target_id: str = Field(min_length=32, max_length=32)
    expected_target_lock_version: int = Field(ge=1)
    condition_text: str = Field(default="", max_length=2_000)
    note: str = Field(default="", max_length=2_000)

    @field_validator("condition_text", "note")
    @classmethod
    def strip_text(cls, value: str) -> str:
        return value.strip()


class ForeshadowPlanCommand(ForeshadowLifecycleCommand):
    expected_structure_version: int = Field(ge=1)
    expected_item_lock_version: int = Field(ge=1)


class ForeshadowFactCreate(ForeshadowLifecycleCommand):
    expected_structure_version: int = Field(ge=1)
    fact_kind: Literal["planted", "resolved"]
    chapter_id: str = Field(min_length=32, max_length=32)
    expected_chapter_lock_version: int = Field(ge=1)
    note: str = Field(default="", max_length=2_000)

    @field_validator("note")
    @classmethod
    def strip_note(cls, value: str) -> str:
        return value.strip()


class ForeshadowFactRetract(ForeshadowLifecycleCommand):
    expected_fact_lock_version: int = Field(ge=1)
    reason: str = Field(min_length=1, max_length=2_000)

    @field_validator("reason")
    @classmethod
    def strip_reason(cls, value: str) -> str:
        if not (clean := value.strip()):
            raise ValueError("撤回原因不能为空")
        return clean


class ForeshadowElementSnapshot(BaseModel):
    id: str
    name: str
    summary: str
    confirmation_status: Literal["candidate", "confirmed", "rejected"]
    lifecycle_status: Literal["active", "archived", "merged"]
    enabled: bool
    content_version: int = Field(ge=1)
    lock_version: int = Field(ge=1)

    model_config = ConfigDict(extra="forbid")


class ForeshadowTargetSnapshot(BaseModel):
    target_type: Literal["part", "chapter"]
    target_id: str
    title: str
    status: Literal["active", "archived"]
    part_id: str | None = None
    position: int = Field(ge=1)

    model_config = ConfigDict(extra="forbid")


class ForeshadowPlanItemResponse(BaseModel):
    id: str
    action_kind: Literal["plant", "resolve"]
    target: ForeshadowTargetSnapshot
    condition_text: str
    note: str
    status: Literal["active", "cancelled"]
    lock_version: int = Field(ge=1)
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(extra="forbid")


class ForeshadowFactResponse(BaseModel):
    id: str
    fact_kind: Literal["planted", "resolved"]
    chapter: ForeshadowTargetSnapshot
    note: str
    status: Literal["active", "retracted"]
    lock_version: int = Field(ge=1)
    created_at: datetime
    retracted_at: datetime | None = None

    model_config = ConfigDict(extra="forbid")


class ForeshadowLifecycleResponse(BaseModel):
    id: str
    project_id: str
    plan_id: str
    status: Literal["active", "archived"]
    state: ForeshadowState
    lock_version: int = Field(ge=1)
    element: ForeshadowElementSnapshot
    plans: list[ForeshadowPlanItemResponse] = Field(default_factory=list)
    facts: list[ForeshadowFactResponse] = Field(default_factory=list)
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(extra="forbid")


class ForeshadowStateCounts(BaseModel):
    unplanted: int = Field(ge=0)
    planted: int = Field(ge=0)
    pending_resolution: int = Field(ge=0)
    resolved: int = Field(ge=0)

    model_config = ConfigDict(extra="forbid")


class ForeshadowLifecycleListResponse(BaseModel):
    items: list[ForeshadowLifecycleResponse]
    counts: ForeshadowStateCounts
    next_cursor: str | None = None

    model_config = ConfigDict(extra="forbid")


class ForeshadowEventResponse(BaseModel):
    id: str
    event_kind: Literal[
        "create",
        "archive",
        "restore",
        "plan_create",
        "plan_cancel",
        "plan_restore",
        "fact_record",
        "fact_retract",
    ]
    plan_item_id: str | None = None
    fact_id: str | None = None
    previous_lifecycle_version: int = Field(ge=0)
    new_lifecycle_version: int = Field(ge=1)
    metadata: dict[str, Any]
    created_at: datetime

    model_config = ConfigDict(extra="forbid")


class ForeshadowHistoryResponse(BaseModel):
    lifecycle_id: str
    items: list[ForeshadowEventResponse]

    model_config = ConfigDict(extra="forbid")


class ForeshadowMutationReceipt(BaseModel):
    receipt_id: str
    operation_key: str
    operation_type: Literal[
        "foreshadow_bind",
        "foreshadow_archive",
        "foreshadow_restore",
        "foreshadow_plan_create",
        "foreshadow_plan_cancel",
        "foreshadow_plan_restore",
        "foreshadow_fact_record",
        "foreshadow_fact_retract",
    ]
    replayed: bool
    project_id: str
    lifecycle_id: str
    previous_lifecycle_version: int = Field(ge=0)
    new_lifecycle_version: int = Field(ge=1)
    event_id: str
    lifecycle: ForeshadowLifecycleResponse
    created_at: datetime

    model_config = ConfigDict(extra="forbid")

File: frontend/src/types/foreshadow.ts
export type ForeshadowState = "unplanted" | "planted" | "pending_resolution" | "resolved";
export type ForeshadowLifecycleStatus = "active" | "archived";
export type ForeshadowActionKind = "plant" | "resolve";
export type ForeshadowFactKind = "planted" | "resolved";

export interface ForeshadowElementSnapshot {
  id: string;
  name: string;
  summary: string;
  confirmation_status: "candidate" | "confirmed" | "rejected";
  lifecycle_status: "active" | "archived" | "merged";
  enabled: boolean;
  content_version: number;
  lock_version: number;
}

export interface ForeshadowTargetSnapshot {
  target_type: "part" | "chapter";
  target_id: string;
  title: string;
  status: "active" | "archived";
  part_id: string | null;
  position: number;
}

export interface ForeshadowPlanItem {
  id: string;
  action_kind: ForeshadowActionKind;
  target: ForeshadowTargetSnapshot;
  condition_text: string;
  note: string;
  status: "active" | "cancelled";
  lock_version: number;
  created_at: string;
  updated_at: string;
}

export interface ForeshadowFact {
  id: string;
  fact_kind: ForeshadowFactKind;
  chapter: ForeshadowTargetSnapshot;
  note: string;
  status: "active" | "retracted";
  lock_version: number;
  created_at: string;
  retracted_at: string | null;
}

export interface ForeshadowLifecycle {
  id: string;
  project_id: string;
  plan_id: string;
  status: ForeshadowLifecycleStatus;
  state: ForeshadowState;
  lock_version: number;
  element: ForeshadowElementSnapshot;
  plans: ForeshadowPlanItem[];
  facts: ForeshadowFact[];
  created_at: string;
  updated_at: string;
}

export interface ForeshadowStateCounts {
  unplanted: number;
  planted: number;
  pending_resolution: number;
  resolved: number;
}

export interface ForeshadowListResponse {
  items: ForeshadowLifecycle[];
  counts: ForeshadowStateCounts;
  next_cursor: string | null;
}

export interface ForeshadowListFilters {
  status?: ForeshadowLifecycleStatus;
  state?: ForeshadowState;
  after?: string;
  limit?: number;
}

export type ForeshadowEventKind =
  | "create"
  | "archive"
  | "restore"
  | "plan_create"
  | "plan_cancel"
  | "plan_restore"
  | "fact_record"
  | "fact_retract";

export interface ForeshadowEvent {
  id: string;
  event_kind: ForeshadowEventKind;
  plan_item_id: string | null;
  fact_id: string | null;
  previous_lifecycle_version: number;
  new_lifecycle_version: number;
  metadata: Record<string, unknown>;
  created_at: string;
}

export interface ForeshadowHistoryResponse {
  lifecycle_id: string;
  items: ForeshadowEvent[];
}

export type ForeshadowOperationType =
  | "foreshadow_bind"
  | "foreshadow_archive"
  | "foreshadow_restore"
  | "foreshadow_plan_create"
  | "foreshadow_plan_cancel"
  | "foreshadow_plan_restore"
  | "foreshadow_fact_record"
  | "foreshadow_fact_retract";

export interface ForeshadowMutationReceipt {
  receipt_id: string;
  operation_key: string;
  operation_type: ForeshadowOperationType;
  replayed: boolean;
  project_id: string;
  lifecycle_id: string;
  previous_lifecycle_version: number;
  new_lifecycle_version: number;
  event_id: string;
  lifecycle: ForeshadowLifecycle;
  created_at: string;
}

export interface ForeshadowBindInput {
  operation_key: string;
  element_id: string;
  expected_structure_version: number;
  expected_element_lock_version: number;
}

export interface ForeshadowLifecycleInput {
  operation_key: string;
  expected_lifecycle_version: number;
}

export interface ForeshadowRestoreInput extends ForeshadowLifecycleInput {
  expected_structure_version: number;
  expected_element_lock_version: number;
}

export interface ForeshadowPlanCreateInput extends ForeshadowLifecycleInput {
  expected_structure_version: number;
  action_kind: ForeshadowActionKind;
  target_type: "part" | "chapter";
  target_id: string;
  expected_target_lock_version: number;
  condition_text: string;
  note: string;
}

export interface ForeshadowPlanStateInput extends ForeshadowLifecycleInput {
  expected_structure_version: number;
  expected_item_lock_version: number;
}

export interface ForeshadowFactCreateInput extends ForeshadowLifecycleInput {
  expected_structure_version: number;
  fact_kind: ForeshadowFactKind;
  chapter_id: string;
  expected_chapter_lock_version: number;
  note: string;
}

export interface ForeshadowFactRetractInput extends ForeshadowLifecycleInput {
  expected_fact_lock_version: number;
  reason: string;
}

export type ForeshadowWritePayload =
  | ForeshadowBindInput
  | ForeshadowLifecycleInput
  | ForeshadowRestoreInput
  | ForeshadowPlanCreateInput
  | ForeshadowPlanStateInput
  | ForeshadowFactCreateInput
  | ForeshadowFactRetractInput;
